network/p2p/unicast/message_rate_limiter.go

```diff
@@ -0,0 +1,89 @@
+package unicast
+
+import (
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/onflow/flow-go/network/message"
+)
+
+// MessageRateLimiterImpl unicast rate limiter that limits the amount of streams that can
+// be created per some configured interval. A new stream is created each time a libP2P
+// node sends a direct message.
+type MessageRateLimiterImpl struct {
+	limiters                 *rateLimiterMap
+	limit                    rate.Limit
+	burst                    int
+	now                      GetTimeNow
+	rateLimitLockoutDuration time.Duration // the amount of time that has to pass before a peer is allowed to connect
+}
+
+// NewMessageRateLimiter returns a new MessageRateLimiterImpl. The cleanup loop will be started in a
+// separate goroutine and should be stopped by calling Close.
+func NewMessageRateLimiter(limit rate.Limit, burst, lockoutDuration int, opts ...RateLimiterOpt) *MessageRateLimiterImpl {
+	l := &MessageRateLimiterImpl{
+		limiters:                 newLimiterMap(rateLimiterTTL, cleanUpTickInterval),
+		limit:                    limit,
+		burst:                    burst,
+		now:                      time.Now,
+		rateLimitLockoutDuration: time.Duration(lockoutDuration) * time.Second,
+	}
+
+	for _, opt := range opts {
+		opt(l)
+	}
+
+	return l
+}
+
+// Allow checks the cached limiter for the peer and returns limiter.Allow().
+// If a limiter is not cached for a peer one is created.
+func (s *MessageRateLimiterImpl) Allow(peerID peer.ID, _ *message.Message) bool {
+	limiter := s.getLimiter(peerID)
+	if !limiter.AllowN(s.now(), 1) {
+		s.limiters.updateLastRateLimit(peerID, s.now())
+		return false
+	} else {
+		return true
+	}
+}
+
+// IsRateLimited returns true is a peer is currently rate limited.
+func (s *MessageRateLimiterImpl) IsRateLimited(peerID peer.ID) bool {
+	metadata, ok := s.limiters.get(peerID)
+	if !ok {
+		return false
+	}
+	return time.Since(metadata.lastRateLimit) < s.rateLimitLockoutDuration
```

Comment:
`Get` is already protected by a lock, the rest of the func only reads so this is indeed concurrency safe.

Reply:
`Get` prevents concurrent reads/writes to the map. Here, we're accessing `metadata.lastRateLimit` outside of that lock. Writes are protected inside `UpdateLastRateLimit`, but reads are not so there's nothing to prevent concurrent reads/writes on `lastRateLimit`.

There are a number of ways to accomplish this. one is atomics like I mentioned [here](https://github.com/onflow/flow-go/pull/2840#discussion_r995332394). You could also add a getter method to `RateLimiterMap` that takes a read lock and returns the last rate limited timestamp. The downside of that is you're taking a global lock instead of one per entry. Alternatively, you could add a mutex on the `RateLimiterMetadata` object itself to protect accesses